File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
def check_max(mid, max_length):
    if mid > max_length:
        max_length = mid
    return max_length


def binary_search(N, lines):
    low, high = 0, 10000000000
    max_length = 0

    while low <= high:
        mid = (low + high) // 2
        length = 0

        for line in lines:
            length += line//mid

        if length >= N:
            low = mid + 1
            max_length = check_max(mid, max_length)

        else:
            high = mid - 1

    return max_length


if __name__ == "__main__":
    K, N = map(int, input().split())
    lines = [int(input()) for _ in range(K)]

    print(binary_search(N, lines))

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import static java.lang.Integer.*;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	public static void main(String[] args) throws IOException {

		BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
		// BufferedReader bufferedReader = new BufferedReader(new FileReader("input.txt"));

		StringTokenizer stringTokenizer = new StringTokenizer(bufferedReader.readLine());
		int lineCounts = parseInt(stringTokenizer.nextToken()); // 1<= lineCounts <=10000
		int linesInNeed = parseInt(stringTokenizer.nextToken()); // 1<= linesInNeed < 1000000
		// lineCounts <= linesInNeed

		int[] lines = new int[lineCounts];
		int maxLineLength = 0;
		while (lineCounts-- > 0) {
			stringTokenizer = new StringTokenizer(bufferedReader.readLine());
			lines[lineCounts] = parseInt(stringTokenizer.nextToken());
			maxLineLength = Math.max(maxLineLength, lines[lineCounts]);
		}

		long low = 1;
		// long high = maxLineLength;
		long high = (1L << 31) - 1;
		long answer = 0;

		while (low <= high) {
			long mid = (low + high) / 2;
			long count = calculateCount(lines, mid);

			if (count >= linesInNeed) {
				answer = mid;
				low = mid + 1;
			} else {
				high = mid - 1;
			}
		}

		System.out.println(answer);
	}

	public static int calculateCount(int[] lengths, long x) {
		int count = 0;
		for (int length : lengths) {
			count += length / x;
		}
		return count;
	}

}
